### DISCUSSION TITLE
Next JS useMediaQuery Hook returns undefined for the the 1st page load, later returns the proper screen size.

### DISCUSSION THREAD

**Discussion by BasavarajPatil24** [vercel/next.js#46466] [Help]
I am using custom hook to measure the screen size and load the pages for the mobile and desktop views in NextJS project. I have created the custom Hook as below which returns boolean value according to the input size given to it as mentioned in the breakpoints.

If we are passing the breakpoint of 576, it returns true for the small screen, same way apply for large screens also.

```
import { useState, useLayoutEffect } from 'react';

export enum Breakpoints {
  'S' = 576,
  'M' = 768,
  'L' = 1024,
  'XL' = 1280,
  'XXL' = 1440,
  'XXXL' = 2280,
}

type useMediaQueryReturn = Array<boolean | number>;

export const useMediaQuery = (breakpoints?: Breakpoints | Breakpoints[]): useMediaQueryReturn => {
  
  const [width, setWidth] = useState(0);
  const [values, setValues] = useState<boolean[]>([]);

  const updateWidth = () => setWidth(window.innerWidth);

  useLayoutEffect(() => {
    updateWidth();
    window.addEventListener('resize', updateWidth);

    return () => window.removeEventListener('resize', updateWidth);
  }, []);

  useLayoutEffect(() => {
    const returnValues = [];

    if (Array.isArray(breakpoints)) {
      breakpoints.forEach((breakpoint) => {
        returnValues.push(window.matchMedia(`(min-width: ${breakpoint}px)`).matches);
      });
    } else if (breakpoints) {
      returnValues.push(window.matchMedia(`(min-width: ${breakpoints}px)`).matches);
    }

    setValues(returnValues);
  }, [width]);

  return breakpoints ? [...values, width] : [width];
};
```

This above code returns `undefined` for the first time, and later it gives the proper value which is resulting in CLS (Cumulative Layout Shift). Is there any way to resolves these issues?

**Comment by icyJoseph** [comment]
Typically the issue with this MediaQuery hooks is that during SSR there's no window, no document, no DOM elements.

This is because SSR happens on another computer, on a server rather. 

So there's no ability at all to measure the screen. Depending on the implementation various things can be returned, but it won't be by any means a proper screen measurement.

With that out of the way.

`useLayoutEffect` is an interesting case, because we can use to change the DOM before browser paint. Since there's no browser in SSR, often use-layout-effect is labelled as a NOOP during SSR. Some people create an `useIsomorphicEffect`, which depending on the rendering context, uses `useEffect` or `useLayoutEffect`.

Last but not least, these effects are triggering a bunch of re-renders that could be optimised away, specially because of use-layout-effect usage, but let's first try to resolve your issue.

Now, onto your problem:

> This above code returns undefined for the first time, and later it gives the proper value which is resulting in CLS (Cumulative Layout Shift). Is there any way to resolves these issues?

I am not sure why would you get `undefined` from:

```js
 return breakpoints ? [...values, width] : [width];
```

This should clearly resolve to an array in both cases. 

I think you probably mean that `values` is empty, but you want to read the `width`, which would be in position `[1]`, because breakpoints is passed, however you read `undefined`, because an empty list spread with `width`, would just be `width`?

Regardless, I think the problem is that you are trying to render for narrow and wide screens, while on the server, and that necessarily generates a first frame with incorrect styles.

It is only later during client side rendering that these hooks have any meaning.

What you could do is render these with CSS visibility, but that will probably still cause layout shifts.

Some might say to use one of these techniques: https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85, but those might hurt your "SEO", since the initial HTML payload, will not have that content. Other forms of this mistake are using a flag that's set client side, etc, again, this means that the content will only be rendered if JavaScript is available.

The right solution here to use is CSS, an alternative would be to render a proper default, perhaps hiding it with CSS visibility, etc etc, in this scenario (having SEO and also proper layout based on client), any approach that's not CSS based will have trade-offs.

--- 

One additional thing, this effect:

```js
  useLayoutEffect(() => {
    const returnValues = [];

    if (Array.isArray(breakpoints)) {
      breakpoints.forEach((breakpoint) => {
        returnValues.push(
          window.matchMedia(`(min-width: ${breakpoint}px)`).matches
        );
      });
    } else if (breakpoints) {
      returnValues.push(
        window.matchMedia(`(min-width: ${breakpoints}px)`).matches
      );
    }

    setValues(returnValues);
  }, [width]);
```

Is missing `breakpoints` as a dependency, I guess you are skipping it because it might cause re-rendering from the consumer side?

**Comment by ciruz** [comment]
I can recommend you this library, https://github.com/yocontra/react-responsive it works well with next.js, like so:

```javascript
export default function Foobar() {
  const [hydrated, setHydrated] = useState(false);

  const isTabletOrMobile = useMediaQuery(
    {
      query: "(max-width: 1224px)",
      //query: "(min-width:0px) and (max-width: 1224px)", // just as another example
    },
    hydrated ? undefined : { deviceWidth: 660 }
  );

  useEffect(() => {
    setHydrated(true);
  }, []);

  return (
   <div>{isTabletOrMobile && "..."}</div>
  );
```
